### PR TITLE
fix(common): Cap processor affinity at the native mask width

### LINKED ISSUE #249: ProcessExtensions affinity mask ignores pointer/mask width (>64 CPUs or 32-bit process)
Surfaced by review bots (Copilot + codeant) during #245. **Pre-existing** on `master` — #245 only added null-guards, `[SupportedOSPlatform]` annotations, and made the existing `(IntPtr)` cast explicitly `unchecked` (no behavioural change; C# casts are unchecked by default). Tracking the underlying bug here.

`src/Common/Diagnostics/ProcessExtensions.cs` — `SetSingleProcessorAffinity`, `SetEnabledProcessors`, `GetEnabledProcessors`.

The affinity bitmask is built with `1L << processorNumber`, and the accepted range is validated only against `Environment.ProcessorCount`. Two edge-hardware defects:

- **>64 logical processors:** `1L << n` for `n >= 64` wraps the shift count (`n & 63`), producing the wrong bit.
- **32-bit process:** `Process.ProcessorAffinity` is a 32-bit `IntPtr`; `unchecked((IntPtr)mask)` silently truncates bits 32–63, so CPUs 32+ become unaddressable and `GetEnabledProcessors` mis-reports.

### Suggested fix
Cap accepted processor numbers to the representable mask width — `IntPtr.Size * 8` — (in addition to `Environment.ProcessorCount`) in all three methods, throwing `ArgumentOutOfRangeException` when exceeded, and bound the `GetEnabledProcessors` loop the same way. Add tests covering the boundary.

### PR DESCRIPTION
**Pull request by kploch**
### Summary

`SetSingleProcessorAffinity`, `SetEnabledProcessors` and `GetEnabledProcessors` in `Ploch.Common.Diagnostics.ProcessExtensions` validated processor numbers only against `Environment.ProcessorCount`, ignoring the width of the native `Process.ProcessorAffinity` bitmask. Two defects (pre-existing on `master`, surfaced by review bots during #245):

- **>64 logical processors:** `1L << n` for `n >= 64` wraps the shift count (`n & 63`), silently setting the wrong affinity bit.
- **32-bit process:** `Process.ProcessorAffinity` is a 32-bit `IntPtr`, so bits 32–63 were silently truncated and `GetEnabledProcessors` mis-reported.

### Changes

- Processor numbers are now validated against the lesser of `Environment.ProcessorCount` and the native affinity-mask width (`IntPtr.Size * 8`); unaddressable numbers throw `ArgumentOutOfRangeException` with a message stating both limits (`src/Common/Diagnostics/ProcessExtensions.cs`).
- `GetEnabledProcessors` bounds its read loop the same way, so it never reports processors that cannot be represented in the mask.
- XML docs updated on all three methods to document the mask-width constraint.
- Change-log entry added (`change-log/2026-07-12-249-processextensions-affinity-mask-width.md`).

### Design decisions

- **Internal, parameter-injected validation helpers** (`GetMaxAddressableProcessors`, `ValidateProcessorNumber`) + `InternalsVisibleTo("Ploch.Common.Tests")`: the defect only manifests on >64-CPU machines or in 32-bit processes, neither of which occurs on dev/CI hardware. Injecting `processorCount`/`affinityMaskWidth` makes both defect scenarios genuinely unit-testable. Reflection-based access was rejected (Sonar S3011); a public helper was rejected (unnecessary API surface).
- **JetBrains.Annotations alias in the test project:** `InternalsVisibleTo` exposed `Ploch.Common`'s internal JetBrains annotations to the test assembly, colliding (CS0433) with the `JetBrains.Annotations` package. The package reference is now aliased (`Aliases="JetBrainsAnnotations"` + `extern alias` in `Reflection/TestTypes.cs`), keeping both deterministically resolvable.
- **Single combined cap** rather than separate errors per limit: the exception message states both limits and the effective bound, which is what a caller needs to correct the input.
- The existing `unchecked((IntPtr)mask)` casts and CA2020 pragmas are retained — still required so bit 31 wraps to the native pointer pattern in a 32-bit process instead of throwing `OverflowException`.

### Behaviour change (not breaking for supported configurations)

Calls that previously applied a silently wrong affinity mask (processor number within `ProcessorCount` but beyond the mask width) now throw `ArgumentOutOfRangeException`. Behaviour is unchanged on ≤64-CPU machines running 64-bit processes.

### Testing

- 14 new test cases in `ProcessExtensionsTests` (boundary tests at `IntPtr.Size * 8` for both setters, previously-missing out-of-range theory for `SetEnabledProcessors`, injected-limit theories covering the >64-CPU and 32-bit-mask scenarios for both internal helpers).
- Full solution: build succeeded with zero new warnings; `Ploch.Common.Tests` 841 passed (net10.0) / 808 passed (net8.0), 0 failed.
- Coverage on the new/changed code paths is complete (every branch of the validation helper is exercised).
- Note: plan/code review via Codex MCP was unavailable (account model restriction); an independent local review agent pass was used instead and its one finding (loop-bound logic not testable with injected limits) was addressed by extracting `GetMaxAddressableProcessors`.

### Review-pass follow-up (commit c495eb3)

- Added three cross-platform (Windows + Linux) tests covering the success paths of `SetSingleProcessorAffinity`, `SetEnabledProcessors` and `GetEnabledProcessors` — previously only exercised by a Windows-gated test, which left new-code coverage at 77.8% on the Linux CI runner (SonarCloud gate threshold: 80%). The setter tests mutate a throwaway child process (`ping`/`sleep`) rather than the test host so parallel xUnit collections are unaffected; the read-back test verifies against the OS-reported affinity mask bits.
- Hoisted `MaxAddressableProcessors` into a local before the `GetEnabledProcessors` loop (review suggestion from gemini-code-assist and codereviewbot-ai).
- Full suite after the change: 838 passed (net10.0) / 805 passed (net8.0), 0 failed; zero build warnings on modified files.

## Issue ticket number and link

Closes #249 — https://github.com/mrploch/ploch-common/issues/249

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? — N/A
- [ ] Will this be part of a product update? If yes, please write one phrase about this update. — Yes: processor-affinity extensions now reject processor numbers that cannot be represented in the native affinity mask instead of silently misbehaving (part of v4.0).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Cap processor-affinity operations to the native affinity-mask width and ensure they reject unaddressable processor numbers while accurately reporting enabled processors.

Bug Fixes:
- Prevent processor-affinity APIs from silently setting or reporting incorrect affinity bits on machines with more than 64 logical processors or in 32-bit processes by capping processor numbers to the native mask width.
- Ensure GetEnabledProcessors does not report processors beyond what can be represented in the native ProcessorAffinity bitmask.

Enhancements:
- Introduce shared internal helpers to compute the maximum addressable processors and validate processor numbers against both processor count and affinity-mask width.
- Clarify XML documentation for processor-affinity extension methods to describe the mask-width constraint and resulting exceptions.

Documentation:
- Add a change-log entry documenting the processor-affinity mask-width fix and its behavioral impact.

Tests:
- Add unit tests covering mask-width boundary conditions, invalid processor numbers, and the new internal validation helpers for various processor-count and mask-width combinations.
- Adjust test project references to JetBrains.Annotations via an assembly alias to avoid type conflicts after exposing internals to tests.

___

## **CodeAnt-AI Description**
**Prevent invalid processor affinity settings on large CPU counts and 32-bit processes**

### What Changed
- Processor affinity now rejects processor numbers that cannot be represented by the current process’s native mask, instead of silently setting or reporting the wrong CPU
- `SetSingleProcessorAffinity`, `SetEnabledProcessors`, and `GetEnabledProcessors` now follow the same limit, so enabled processors are only returned when they can actually be addressed
- Tests now cover out-of-range processor numbers, mask-width edge cases, and the test project avoids a type-name clash caused by the new internal access
- Documentation and change-log notes now explain the mask-width limit and the resulting behavior change

### Impact
`✅ Fewer wrong affinity settings on servers with many CPUs`
`✅ Correct processor lists in 32-bit processes`
`✅ Clearer errors when a requested CPU cannot be addressed`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved processor-affinity validation to respect both the available processor count and the system’s native affinity-mask width.
  * Invalid or unrepresentable processor numbers now consistently raise an appropriate range exception (with clearer bounds details).
  * Processor enumeration no longer reports processors that cannot be represented by the native affinity mask.
* **Tests**
  * Added coverage for processor-affinity mask-width validation scenarios.
* **Documentation**
  * Updated API documentation and added a changelog entry describing the corrected affinity behavior.

### CODE REVIEW

File: change-log/2026-07-12-249-processextensions-affinity-mask-width.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+- `GetEnabledProcessors` bounds its read loop the same way, so it never reports
+  processors that cannot be represented in the mask.
+
+## Behaviour change (not breaking for supported configurations)
+
+Previously, on hardware/process combinations where a processor number passed the
+`ProcessorCount` check but exceeded the mask width, the methods silently applied a
+wrong affinity mask. These calls now throw `ArgumentOutOfRangeException`. On machines
+with ≤64 logical processors running 64-bit processes (the overwhelmingly common case)
+behaviour is unchanged.
```
[the remark this answers is not in the document]
Comment (kploch) [reply change-log/2026-07-12-249-processextensions-affinity-mask-width.md:28]: I believe this is a false positive. The custom rule targets persistent *decision/lesson notes* (the kind that belong in ContextStream), but `change-log/*.md` entries are a different, mandated artefact in this repository:

- The repo rule `.claude/rules/change-log.md` explicitly **requires** a change-log entry per task ("When working on a task, and before doing a commit, create a file describing the change"), with this exact filename format (`<date>-#<taskid>-<name>.md`).
- The release pipeline consumes these files: `.github/workflows/release.yml` generates the GitHub Release notes from `change-log/*.md` and then archives consumed entries to `change-log/archive/` (see also CLAUDE.md § Release Pipeline, steps 6–8).

Removing this file would break the documented release-notes workflow for the v4.0 release. The behaviour-change description it contains is release-notes material, not a session lesson. Resolving as false positive.

File: src/Common/Diagnostics/ProcessExtensions.cs [hunk 2]
```diff
@@ -95,7 +126,7 @@ public static IEnumerable<int> GetEnabledProcessors(this Process process)
         // call time instead of being deferred until the sequence is enumerated.
         var affinityMask = process.ProcessorAffinity.ToInt64();
         var enabledProcessors = new List<int>();
-        for (var i = 0; i < Environment.ProcessorCount; i++)
+        for (var i = 0; i < MaxAddressableProcessors; i++)
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs]: Done in c495eb3 — `MaxAddressableProcessors` is now read once into a local (`maxAddressableProcessors`) before the loop and used as the bound. Thanks!

[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs]: Done in c495eb3 — `MaxAddressableProcessors` is now read once into a local (`maxAddressableProcessors`) before the loop and used as the bound, exactly as suggested. Thanks!

File: src/Common/Diagnostics/ProcessExtensions.cs [hunk 3]
```diff
@@ -105,4 +137,41 @@ public static IEnumerable<int> GetEnabledProcessors(this Process process)
...
+    /// <paramref name="processorCount"/> and <paramref name="affinityMaskWidth"/>.
+    /// </summary>
+    /// <param name="processorCount">The number of logical processors on the machine.</param>
+    /// <param name="affinityMaskWidth">The width, in bits, of the native affinity mask.</param>
+    /// <returns>The exclusive upper bound for addressable processor numbers.</returns>
+    /// <remarks>
+    /// Internal (rather than private) so the mask-width capping can be unit-tested with limits that do not occur on the
+    /// test machine — more than 64 logical processors, or the 32-bit-process mask width.
+    /// </remarks>
+    internal static int GetMaxAddressableProcessors(int processorCount, int affinityMaskWidth) => Math.Min(processorCount, affinityMaskWidth);
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Diagnostics/ProcessExtensions.cs:152]: Agreed the observation is technically correct: since .NET 6, `Environment.ProcessorCount` is the number of CPUs *available to the process* (it already respects the process's own affinity and CPU limits), so it is a count, not an upper bound on valid CPU indices — a process constrained to a non-contiguous set such as CPUs 8–15 reports `ProcessorCount == 8`, and both the loop bound and the validators would mis-handle indices 8–15.

However, this is **pre-existing behaviour on `master`**: the `Environment.ProcessorCount` bound existed in all three methods before this PR (`git log -S "Environment.ProcessorCount" -- src/Common/Diagnostics/ProcessExtensions.cs`). This PR deliberately scoped itself to the mask-width wrap/truncation defects from #249 and documented the combined `min(ProcessorCount, mask width)` cap as its validation contract (see the "Single combined cap" design decision in the PR description and the XML docs). Changing the setters' validation contract for constrained-affinity processes is a behaviour/design change that deserves its own review and constrained-affinity test scenarios rather than a late amendment to this PR.

Deferred to follow-up issue **#257** (https://github.com/mrploch/ploch-common/issues/257), which captures your suggested direction: bound `GetEnabledProcessors` by the mask width alone, and make a deliberate decision on the setter validation contract. Resolving this thread as out-of-scope-with-follow-up.

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs [hunk 4]
```diff
@@ -76,11 +87,135 @@ public void SetEnabledProcessors_should_throw_when_no_processor_numbers_are_spec
...
+    public void SetSingleProcessorAffinity_should_enable_only_the_requested_processor()
+    {
+        // A throwaway child process is mutated (never the test host) so parallel test collections are unaffected.
+        using var childProcess = StartShortLivedChildProcess();
+
+        try
+        {
+            childProcess.SetSingleProcessorAffinity(0);
+
+            childProcess.GetEnabledProcessors().Should().Equal(0);
```
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs]: Done in 800782e — both setter tests now derive their target processors from the child's OS-reported affinity mask via a `GetAllowedProcessors` helper (raw `ProcessorAffinity.ToInt64()` bits, bounded by `min(ProcessorCount, IntPtr.Size * 8)`), so they no longer assume CPU 0 is in the allowed set under cpuset/container CPU pinning. Thanks!

File: tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs [hunk 5]
```diff
@@ -76,11 +87,135 @@ public void SetEnabledProcessors_should_throw_when_no_processor_numbers_are_spec
...
+        }
+        finally
+        {
+            childProcess.Kill();
+        }
```
[the remark this answers is not in the document]
Comment (kploch) [reply tests/Common.Tests/Diagnostics/ProcessExtensionsTests.cs:153]: Done in 800782e — cleanup now goes through a `KillAndReap` helper: `Kill()` is guarded with `HasExited`, and `WaitForExit()` reaps the child, so an exit race can't throw from the `finally` block or mask a real assertion failure. Thanks!
